p2p/src/utils/requester.rs

```diff
@@ -0,0 +1,348 @@
+//! Make concurrent requests to peers limited by rate and prioritized by performance.
+
+use commonware_cryptography::{PublicKey, Scheme};
+use commonware_runtime::Clock;
+use commonware_utils::PrioritySet;
+use governor::{
+    clock::Clock as GClock, middleware::NoOpMiddleware, state::keyed::HashMapStateStore, Quota,
+    RateLimiter,
+};
+use std::{
+    collections::{BTreeMap, BTreeSet, HashMap, HashSet},
+    time::{Duration, SystemTime},
+};
+
+/// Unique identifier for a request.
+///
+/// Once u64 requests have been made, the ID wraps around (resetting to zero).
+/// As long as there are less than u64 requests outstanding, this should not be
+/// an issue.
+pub type ID = u64;
+
+/// Configuration for `Requester`.
+pub struct Config<C: Scheme> {
+    /// Cryptographic primitives.
+    pub crypto: C,
+
+    /// Rate limit for requests per participant.
+    pub rate_limit: Quota,
+
+    /// Initial expected performance for new participants.
+    pub initial: Duration,
+
+    /// Timeout for requests.
+    pub timeout: Duration,
+}
+
+/// Send rate-limited requests to peers prioritized by performance.
+pub struct Requester<E: Clock + GClock, C: Scheme> {
+    runtime: E,
+    crypto: C,
+    initial: Duration,
+    timeout: Duration,
+
+    rate_limiter:
+        RateLimiter<PublicKey, HashMapStateStore<PublicKey>, E, NoOpMiddleware<E::Instant>>,
+    participants: PrioritySet<PublicKey, u128>,
+    excluded: HashSet<PublicKey>,
+
+    id: ID,
+    requests: HashMap<ID, (PublicKey, SystemTime, SystemTime)>,
+    deadlines: BTreeMap<SystemTime, BTreeSet<ID>>,
```

Comment:
We don't expect a lot of overlapping deadlines (i.e. equal `SystemTime` values). Would it be simpler and more efficient to have a data structure of a single "layer"? That is, simply a sorted set of tuples `(SystemTime, ID)` which is sorted by time then id? Otherwise you need this sorted set of sets, which seems less efficient when we expect every inner set to have 1 value (rarely more).

Reply:
Yeah that's a good point. I'll use the same technique that I employ with `BTreeSet` in `priority_set`.